### PR TITLE
Add basic ArchMath

### PR DESCRIPTION
**Pull request by Twarug**

### CODE REVIEW

File: include/Ref.h [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+template<typename T>
+using Ref = std::shared_ptr<T>;
+
+template<typename T, typename... Args>
+constexpr Ref<T> createRef(Args&&... args) {
+	return std::make_shared<T>(std::forward<Args>(args)...);
+}
+
+} // namespace arch
```
Comment (Tosiekdev) [inline include/Ref.h:15]: newline

Comment (Twarug) [reply include/Ref.h]: sory, this file, shouldn't be in this PR

File: include/math/Math.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+using i32 = int32_t;
+using i64 = int64_t;
+
+using u8 = uint8_t;
+using u16 = uint16_t;
+using u32 = uint32_t;
+using u64 = uint64_t;
+
+using f32 = float_t;
+using f64 = double_t;
```
Comment (Tosiekdev) [inline include/math/Math.h:25]: maybe some alias for long double?

File: src/Transform.cpp [hunk 3]
```diff
@@ -1,59 +1,61 @@
...
+	auto norm = static_cast<f32>(sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2)));
 	x /= norm;
 	y /= norm;
 	z /= norm;
 
 	// fix angle
 	angle /= 2;
-	const double pi = std::acos(-1);
-	angle *= static_cast<float>(pi / 180.0);
+	const f64 pi = std::acos(-1);
```
Comment (Tosiekdev) [inline src/Transform.cpp:16]: pi can be constexpr (gcc users will benefit from this), also you can use std::acosf and get rid of unnecessary conversion

File: src/Transform.cpp [hunk 4]
```diff
@@ -1,59 +1,61 @@
...
-	float c = z * std::sin(angle);
-	glm::vec3 imaginaryPart = glm::vec3(a, b, c);
-	glm::vec4 q = glm::vec4(imaginaryPart, std::cos(angle));
-	glm::vec4 q1 = glm::vec4(-imaginaryPart, std::cos(angle));
-
-	return {q, q1};
+	f32 a = x * std::sin(angle);
+	f32 b = y * std::sin(angle);
+	f32 c = z * std::sin(angle);
+	float3 imaginaryPart = float3(a, b, c);
```
Comment (Tosiekdev) [inline src/Transform.cpp:23]: imaginaryPart{a,b,c} won't work?

File: src/Transform.cpp [hunk 5]
```diff
@@ -1,59 +1,61 @@
...
+	float4 q = float4(imaginaryPart, std::cos(angle));
```
Comment (Tosiekdev) [inline src/Transform.cpp:24]: similar as above

**Review by Tosiekdev — COMMENTED** [review]
Please also check comments not within the review. Moreover, consider using construction with brackets and not by copy constructor.

File: src/Transform.cpp (hunk 3 above)
Comment (Twarug) [reply src/Transform.cpp:16]: TBH i don't think, that this class will stay for long. Probably it'll be removed during Vulkan API implementation

Comment (Tosiekdev) [reply src/Transform.cpp:16]: I think we should prevent this code from situation like // Temporary solution -> committed 5 years ago

File: include/math/Math.h (hunk 2 above)
Comment (Twarug) [reply include/math/Math.h:25]: Problem with it, is that C++20 doesn't defines bit size for `long double`. To do so we would need C++23.
[Fixed width floating-point types](https://en.cppreference.com/w/cpp/types/floating-point)

File: src/Transform.cpp (hunk 3 above)
Comment (Twarug) [reply src/Transform.cpp:16]: I that case i think the best solution it to remove a file.

File: include/math/Math.h (hunk 2 above)
Comment (Tosiekdev) [reply include/math/Math.h:25]: Maybe at least some shortcut for long double then? just for the sake of having nice aliases for most of the types

Comment (Twarug) [reply include/math/Math.h:25]: so what you propose as an alias?

Comment (Tosiekdev) [reply include/math/Math.h:25]: I think simply first letters of long double -> ld

File: src/Transform.cpp (hunk 3 above)
Comment (Tosiekdev) [reply src/Transform.cpp:16]: If you are sure it won't be necessary then please, do it

**Review by Tosiekdev — APPROVED** [review]
Well done!

**Pull request merged**